File: src/application/useCases/resumeStagedBatch.ts
import type { Id, ImportBatch } from '../../domain/entities/types'
import type {
  ImportBatchRepository,
  SourceRecordRepository,
  TransactionRepository,
} from '../ports/repositories'

/**
 * ResumeStagedBatch — تنظيف دفعة انقطعت قبل اكتمالها.
 *
 * يحل قيد ARCHITECTURE.md §10.5: معاملة Firestore لا تغطي مئات المستندات،
 * فالذرّية الحقيقية هناك مستحيلة بمعاملة واحدة.
 *
 * الحل المطبَّق بدلها — **بروتوكول علامة الالتزام**:
 *
 *   ١. تُكتب الدفعة بحالة `staged` **قبل** أي عملية
 *   ٢. تُكتب العمليات وسجلات المصدر (قد تنقطع هنا)
 *   ٣. تتحول الدفعة إلى `committed` بكتابة **واحدة** ذرّية بطبيعتها
 *
 * فأي انقطاع يترك دفعة `staged`، وما كُتب تحتها **غير معتمد**:
 * - `findByFileHash` لا يرى إلا `committed`، فلا تُحسب كاستيراد سابق
 * - هذه الحالة تنظّف ما تحت الدفعة المعلّقة عند فتح التطبيق
 *
 * النتيجة: «صفر أو كامل الدفعة» تتحقق **بالأثر** لا بالمعاملة —
 * لا يظهر للمستخدم أبدًا نصف دفعة معتمدة.
 */

export interface StagedCleanupOutcome {
  batchId: Id
  fileName: string
  deletedTransactions: number
  deletedRecords: number
  /** سبب تعذر التنظيف، لو تعذر. الدفعة تبقى معلّقة ولا تُحذف منها حاجة. */
  error?: string
}

export interface ResumeStagedBatchDeps {
  txns: TransactionRepository
  sources: SourceRecordRepository
  batches: ImportBatchRepository
  /** كم دفعة أخيرة تُفحص. القراءة محدودة — ARCHITECTURE.md §5.6. */
  scanLimit?: number
}

export function makeResumeStagedBatch(deps: ResumeStagedBatchDeps) {
  /** يعرض الدفعات المعلّقة بلا حذف — للعرض على المستخدم قبل التنظيف. */
  async function findStaged(): Promise<ImportBatch[]> {
    const recent = await deps.batches.listRecent(deps.scanLimit ?? 20)
    return recent.filter((b) => b.state === 'staged')
  }

  /**
   * ينظّف دفعة معلّقة: يحذف ما كُتب تحتها ويعلّمها `reverted`.
   *
   * آمن التكرار: لو انقطع التنظيف نفسه تبقى الدفعة `staged` فيُعاد التنظيف
   * في المرة الجاية، وحذف ما هو محذوف أصلًا لا يضر.
   */
  async function cleanup(batchId: Id): Promise<StagedCleanupOutcome> {
    const batch = await deps.batches.findById(batchId)
    if (!batch) throw new Error(`دفعة غير موجودة: ${batchId}`)
    if (batch.state !== 'staged') {
      throw new Error(
        `الدفعة دي حالتها «${batch.state}» مش «staged». التنظيف بيشتغل على المعلّق بس.`,
      )
    }

    const records = await deps.sources.listByBatch(batchId)
    const txnIds = records
      .map((r) => r.transactionId)
      .filter((id): id is Id => id !== null)

    // الترتيب مقصود: العمليات أولًا ثم السجلات ثم علامة الدفعة.
    // لو انقطع بينهما تبقى الدفعة staged ويُعاد التنظيف بلا ضرر.
    if (txnIds.length > 0) await deps.txns.deleteMany(txnIds)
    if (records.length > 0) await deps.sources.deleteMany(records.map((r) => r.id))
    await deps.batches.updateState(batchId, 'reverted')

    return {
      batchId,
      fileName: batch.fileName,
      deletedTransactions: txnIds.length,
      deletedRecords: records.length,
    }
  }

  /**
   * ينظّف كل ما هو معلّق. يُستدعى عند فتح التطبيق.
   *
   * ⚠️ دفعة واحدة تعذر تنظيفها **لا توقف الباقي ولا تفتح التطبيق على خطأ**:
   * عطل sanitize القديم ترك حقل `id` مختلفًا عن مسار الوثيقة، فكان
   * `findById` يرمي فتتوقف كل الشاشات (بلاغ المالك 2026-09-11،
   * tests/acceptance/importVisibility.test.ts).
   */
  async function cleanupAll(): Promise<StagedCleanupOutcome[]> {
    const staged = await findStaged()
    const outcomes: StagedCleanupOutcome[] = []
    for (const batch of staged) {
      try {
        outcomes.push(await cleanup(batch.id))
      } catch (error) {
        outcomes.push({
          batchId: batch.id,
          fileName: batch.fileName,
          deletedTransactions: 0,
          deletedRecords: 0,
          error: error instanceof Error ? error.message : String(error),
        })
      }
    }
    return outcomes
  }

  return { findStaged, cleanup, cleanupAll }
}

File: native-app/app/src/commonMain/kotlin/app/masroufy/usecase/ResumeStagedBatch.kt
package app.masroufy.usecase

import app.masroufy.core.Id
import app.masroufy.core.ImportBatch
import app.masroufy.core.ImportBatchState
import app.masroufy.port.ImportBatchRepository
import app.masroufy.port.SourceRecordRepository
import app.masroufy.port.TransactionRepository

/**
 * ResumeStagedBatch — نقل `resumeStagedBatch.ts`: تنظيف دفعة انقطعت قبل اكتمالها.
 * أي انقطاع بيسيب دفعة `staged`، واللي اتكتب تحتها **غير معتمد** — الحالة دي بتنظّفه
 * عند فتح التطبيق، فـ«صفر أو كامل الدفعة» بتتحقق **بالأثر** (ARCHITECTURE §11.1).
 */

data class StagedCleanupOutcome(
    val batchId: Id,
    val fileName: String,
    val deletedTransactions: Int,
    val deletedRecords: Int,
    /** سبب تعذر التنظيف، لو اتعذر. الدفعة بتفضل معلّقة ومفيش حاجة بتتحذف منها. */
    val error: String? = null,
)

data class ResumeStagedBatchDeps(
    val txns: TransactionRepository,
    val sources: SourceRecordRepository,
    val batches: ImportBatchRepository,
    /** كام دفعة أخيرة بتتفحص. القراية محدودة — ARCHITECTURE §5.6. */
    val scanLimit: Int? = null,
)

class ResumeStagedBatch(private val deps: ResumeStagedBatchDeps) {
    /** بيعرض الدفعات المعلّقة من غير حذف — للعرض على المستخدم قبل التنظيف. */
    suspend fun findStaged(): List<ImportBatch> =
        deps.batches.listRecent(deps.scanLimit ?: 20).filter { it.state == ImportBatchState.STAGED }

    /**
     * بينظّف دفعة معلّقة: بيحذف اللي اتكتب تحتها وبيعلّمها `reverted`.
     * آمن التكرار: لو التنظيف نفسه اتقطع، الدفعة بتفضل `staged` وبيتعاد التنظيف بلا ضرر.
     */
    suspend fun cleanup(batchId: Id): StagedCleanupOutcome {
        val batch = deps.batches.findById(batchId) ?: throw IllegalStateException("دفعة غير موجودة: $batchId")
        if (batch.state != ImportBatchState.STAGED) {
            throw IllegalStateException("الدفعة دي حالتها «${batch.state.wire}» مش «staged». التنظيف بيشتغل على المعلّق بس.")
        }

        val records = deps.sources.listByBatch(batchId)
        val txnIds = records.mapNotNull { it.transactionId }

        // الترتيب مقصود: العمليات الأول وبعدها السجلات وبعدها علامة الدفعة —
        // لو اتقطع بينهم الدفعة بتفضل staged ويتعاد التنظيف بلا ضرر
        if (txnIds.isNotEmpty()) deps.txns.deleteMany(txnIds)
        if (records.isNotEmpty()) deps.sources.deleteMany(records.map { it.id })
        deps.batches.updateState(batchId, ImportBatchState.REVERTED)

        return StagedCleanupOutcome(batchId, batch.fileName, deletedTransactions = txnIds.size, deletedRecords = records.size)
    }

    /**
     * بينظّف كل المعلّق — بيتندى عند فتح التطبيق.
     * ⚠️ دفعة واحدة اتعذر تنظيفها **ما بتوقفش الباقي ولا بتفتح التطبيق على خطأ**
     * (بلاغ المالك 2026-09-11 — `importVisibility.test.ts` في التطبيق الحالي).
     */
    suspend fun cleanupAll(): List<StagedCleanupOutcome> {
        val staged = findStaged()
        return staged.map { batch ->
            try {
                cleanup(batch.id)
            } catch (error: Throwable) {
                StagedCleanupOutcome(batch.id, batch.fileName, deletedTransactions = 0, deletedRecords = 0, error = error.message ?: error.toString())
            }
        }
    }
}
